File: src/7.BacktrackingAlgorithms/_exercise/bai4.ts
// [Quay lui-Nhánh cận]. Bài 4. Dãy con có tổng bằng K
// Cho dãy số A[] = (a1, a2, .., an) và số tự nhiên K. Hãy đưa ra tất cả các dãy con
// của dãy số A[] sao cho tổng các phần tử của dãy con đó đúng bằng K. Các phần
// tử của dãy số A[] được giả thuyết là nguyên dương và không có các phần tử giống
// nhau. Ví dụ với dãy con A[] = {5, 10, 15, 20, 25}, K = 50 ta có 3 dãy con {5, 10,
// 15, 20}, {5, 20, 25}, {10, 15, 25}.

function bai4(arr: number[], k: number): void {
  let X: number[] = new Array();
  arr.sort((a, b) => a - b);

  function Try(i: number, start: number, sum: number) {
    for (let j = start; j < arr.length; j++) {
      X[i] = arr[j];
      if (sum + arr[j] === k) {
        console.log("[", X.slice(1, i + 1).join(" "), "]");
      } else if (sum + arr[j] < k) {
        Try(i + 1, j + 1, sum + arr[j]);
      }
    }
  }

  Try(1, 0, 0);
}

// Test
bai4([5, 10, 15, 20, 25], 50);

File: src/6.GreedyAlgorithms/bai4.cpp
// Cho mảng A[] gồm N số nguyên không âm và số K. Nhiệm vụ của bạn là hãy chia
// mảng A[] thành hai mảng con có kích cỡ K và N-K sao cho hiệu giữa tổng hai
// mảng con là lớn nhất. Ví dụ với mảng A[] = {8, 4, 5, 2, 10}, K=2 ta có kết quả là
// 17 vì mảng A[] được chia thành hai mảng {4, 2} và { 8, 5,10} có hiệu của hai
// mảng con là 23-6=17 là lớn nhất.

#include <iostream>
#include <algorithm>
using namespace std;
typedef long long ll;
int main()
{
    int n, k;
    cin >> n >> k;
    int a[n];
    for (int &x : a)
    {
        cin >> x;
    }
    sort(a, a + n);
    k = min(k, n - k);
    ll s1 = 0, s2 = 0;
    for (int i = 0; i < n; i++)
    {
        if (i < k)
            s1 += a[i];
        else
            s2 += a[i];
    }
    cout << s2 - s1;

    return 0;
}